### DISCUSSION TITLE
Check if Value Exists in CCD (XML)

### DISCUSSION THREAD

**Discussion by nickboxx** [nextgenhealthcare/connect#5734] [Q&A] [answered]
I am looking for the most efficient way of determining whether or not something exists in an XML document. I am receiving CCDs where I get a @code is null type of error for this path:

msg['component']['structuredBody']['component'][i]['section']['entry'][x]['act']['entryRelationship'][0]['observation']['participant']['participantRole']['playingEntity']['code']['@code']

And it is because once you get down to that code that it looks like this:

<code nullFlavor="NA"/>

Trying to get around this issue I used an if statement like this (essentially saying if the path above is not null and not a blank string):

if(msg['component']['structuredBody']['component'][i]['section']['entry'][x]['act']['entryRelationship'][0]['observation']['participant']['participantRole']['playingEntity']['code']['@code'].toString() != null **&&** msg['component']['structuredBody']['component'][i]['section']['entry'][x]['act']['entryRelationship'][0]['observation']['participant']['participantRole']['playingEntity']['code']['@code'].toString() != ''){

But this if statement does not stop the code from erroring, it is instead causing the error. It only seems to work if I do two separate if statements inside of each other with the != '' if statement inside of the != null if statement. Is there a cleaner way of doing this inside of 1 if statement rather than having to use two?

**Comment by jonbartels** [comment]
```
var mycode = msg['component']['structuredBody']['component'][i]['section']['entry'][x]['act']['entryRelationship'][0]['observation']['participant']['participantRole']['playingEntity']['code']['@code'].toString();

if(Packages.org.apache.commons.lang3.StringUtils.isNotBlank(mycode)){
 ... do stuff...
}
```

**Comment by tonygermano** [comment, marked as answer]
When you are working with xml objects, accessing a "property" as you would with a normal javascript object is actually doing a child operation on the xml object.

So, where `msg` is an xml object, the following are all equivalent:
```js
msg['component']
msg.component
msg.child('component')
```

This operation returns an XMLList of all child elements whose name matches the identifier. You can use the `.length()` method to check to see how many child elements are in the evaluated List. If you call a child operation on an empty list, it will return another empty list (which is different than a normal javascript property access, which will return `undefined` when a property is not defined.)

However, when you access a list by a numeric index, it will return that specific element from the list, or `undefined` if you attempt to access past the length of the list. That is where you can get into trouble when trying to access child elements of `undefined`, because you are no longer working with an empty XMLList object. Since you have several places where you are accessing by index, those are all potential places where you can get `undefined` when something does not exist.

You *could* do an undefined check after each index access before you access child elements, but the easiest thing to do is just wrap your assignment in a try/catch block, so that you can assume something in your path does not exist when an exception is thrown. I believe this is what mirth itself does if you were to use a mapper or message builder step created through the GUI.
